File: tests/TestSparkCluster.scala
package org.openbiocure.spark

import org.apache.spark.sql.{SparkSession, DataFrame}
import org.apache.spark.sql.types.{StructType, StructField, StringType, IntegerType}
import org.apache.spark.sql.functions._
import io.delta.tables._
import scala.util.{Try, Success, Failure}
import org.apache.spark.internal.Logging

object TestSparkCluster extends Logging {
  private var spark: SparkSession = _

  def main(args: Array[String]): Unit = {
    logInfo("Starting Spark Cluster Tests...")
    
    try {
      spark = createSparkSession()
      
      val tests = Seq(
        ("Basic Spark", () => testBasicSpark(spark)),
        ("MinIO", () => testMinioConnectivity(spark)),
        ("Hive", () => testHiveMetastore(spark)),
        ("Delta Lake", () => testDeltaLake(spark))
      )
      
      var allPassed = true
      for ((testName, testFunc) <- tests) {
        try {
          logInfo(s"\nRunning $testName test...")
          if (testFunc()) {
            logInfo(s"✓ $testName test passed")
          } else {
            logError(s"✗ $testName test failed")
            allPassed = false
          }
        } catch {
          case e: Exception =>
            logError(s"✗ $testName test failed: ${e.getMessage}")
            allPassed = false
        }
      }
      
      logInfo("\n=== Test Summary ===")
      for ((testName, _) <- tests) {
        val status = if (allPassed) "PASSED" else "FAILED"
        logInfo(s"$testName: $status")
      }
      
      sys.exit(if (allPassed) 0 else 1)
      
    } catch {
      case e: Exception =>
        logError(s"Error during test execution: ${e.getMessage}")
        sys.exit(1)
    } finally {
      if (spark != null) {
        spark.stop()
      }
    }
  }

  def verifyDeltaJars(): Unit = {
    logInfo("=== Verifying Delta Lake JARs ===")
    
    // For local development, skip jar verification
    if (sys.env.getOrElse("SPARK_MASTER_URL", "").startsWith("local")) {
      logInfo("Running in local mode, skipping Delta JAR verification")
      return
    }
    
    // Only verify jars in cluster mode
    val jarPath = "/opt/spark/jars"
    if (!new java.io.File(jarPath).exists()) {
      logInfo(s"Jar path $jarPath does not exist, skipping verification")
      return
    }
    
    try {
      val deltaJars = new java.io.File(jarPath).listFiles()
        .filter(_.getName.startsWith("delta-"))
        .map(_.getName)
      
      logInfo(s"Found Delta JARs in $jarPath:")
      deltaJars.foreach(jar => logInfo(s"- $jar"))
      
      if (deltaJars.isEmpty) {
        throw new RuntimeException("No Delta Lake JARs found in /opt/spark/jars/")
      }
    } catch {
      case e: Exception =>
        logError(s"Error verifying Delta JARs: ${e.getMessage}")
        throw e
    }
    
    logInfo("\nClasspath environment variables:")
    logInfo(s"SPARK_CLASSPATH: ${sys.env.getOrElse("SPARK_CLASSPATH", "Not set")}")
    logInfo(s"SPARK_HOME: ${sys.env.getOrElse("SPARK_HOME", "Not set")}")
    logInfo(s"SPARK_DAEMON_JAVA_OPTS: ${sys.env.getOrElse("SPARK_DAEMON_JAVA_OPTS", "Not set")}")
  }

  def initializeHiveMetastore(): Unit = {
    logInfo("\n=== Initializing Hive Metastore Schema ===")
    
    try {
      logInfo("Initializing Hive metastore schema...")
      // Run schematool to initialize the schema
      val schematoolCmd = Array(
        "/opt/spark/bin/schematool",
        "-dbType", "postgres",
        "-initSchema",
        "-verbose"
      )
      
      val schematoolProcess = Runtime.getRuntime.exec(schematoolCmd)
      val schematoolOutput = scala.io.Source.fromInputStream(schematoolProcess.getInputStream).mkString
      val schematoolError = scala.io.Source.fromInputStream(schematoolProcess.getErrorStream).mkString
      val schematoolExitCode = schematoolProcess.waitFor()
      
      if (schematoolExitCode == 0) {
        logInfo("Successfully initialized Hive metastore schema")
        logInfo(schematoolOutput)
      } else {
        logError(s"Failed to initialize Hive metastore schema: $schematoolError")
        throw new RuntimeException("Failed to initialize Hive metastore schema")
      }
    } catch {
      case e: Exception =>
        logError(s"Error initializing Hive metastore: ${e.getMessage}")
        throw e
    }
  }

  def createSparkSession(): SparkSession = {
    verifyDeltaJars()
    
    val isLocalMode = sys.env.getOrElse("SPARK_MASTER_URL", "").startsWith("local")
    logInfo(s"\nRunning in ${if (isLocalMode) "local" else "cluster"} mode")
    
    if (!isLocalMode) {
      initializeHiveMetastore()
      createMinioBucket()
    }
    
    logInfo("\nSetting up Spark session with debug logging...")
    
    val builder = SparkSession.builder()
      .appName("SparkClusterTest")
      .master(sys.env.getOrElse("SPARK_MASTER_URL", "local[*]"))
      .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
      .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
      .config("spark.driver.log.level", "DEBUG")
      .config("spark.executor.log.level", "DEBUG")
      // Always include MinIO and PostgreSQL configs
      .config("spark.hadoop.fs.s3a.endpoint", sys.env("AWS_ENDPOINT_URL"))
      .config("spark.hadoop.fs.s3a.access.key", sys.env("AWS_ACCESS_KEY_ID"))
      .config("spark.hadoop.fs.s3a.secret.key", sys.env("AWS_SECRET_ACCESS_KEY"))
      .config("spark.hadoop.fs.s3a.path.style.access", "true")
      .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
      .config("spark.hadoop.fs.s3a.aws.credentials.provider", "org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider")
      .config("spark.sql.warehouse.dir", s"s3a://${sys.env("MINIO_BUCKET")}/warehouse")
      .config("spark.sql.catalogImplementation", "hive")
      .config("spark.hadoop.javax.jdo.option.ConnectionURL", s"jdbc:postgresql://${sys.env("POSTGRES_HOST")}:${sys.env("POSTGRES_PORT")}/${sys.env("POSTGRES_DB")}")
      .config("spark.hadoop.javax.jdo.option.ConnectionDriverName", "org.postgresql.Driver")
      .config("spark.hadoop.javax.jdo.option.ConnectionUserName", sys.env("POSTGRES_USER"))
      .config("spark.hadoop.javax.jdo.option.ConnectionPassword", sys.env("POSTGRES_PASSWORD"))
    
    if (isLocalMode) {
      builder
        .config("spark.driver.extraJavaOptions", "-Djava.security.manager=disallow")
        .config("spark.executor.extraJavaOptions", "-Djava.security.manager=disallow")
        .config("spark.local.ip", "127.0.0.1")
        .config("spark.driver.host", "127.0.0.1")
        .config("spark.driver.bindAddress", "127.0.0.1")
    }
    
    builder.getOrCreate()
  }

  def testBasicSpark(spark: SparkSession): Boolean = {
    logInfo("\n=== Testing Basic Spark Functionality ===")
    try {
      val df = createTestDataFrame(spark)
      logInfo("Test DataFrame:")
      df.show()
      true
    } catch {
      case e: Exception =>
        logError(s"Basic Spark test failed: ${e.getMessage}")
        logError(s"Stack trace: ${e.getStackTrace.mkString("\n")}")
        false
    }
  }

  def testMinioConnectivity(spark: SparkSession): Boolean = {
    logInfo("\n=== Testing MinIO Connectivity ===")
    try {
      val df = createTestDataFrame(spark)
      val minioPath = s"s3a://${sys.env("MINIO_BUCKET")}/test/minio_test"
      
      logInfo(s"Writing test data to $minioPath")
      df.write
        .mode("overwrite")
        .parquet(minioPath)
      
      val readDf = spark.read.parquet(minioPath)
      logInfo("Read DataFrame from MinIO:")
      readDf.show()
      true
    } catch {
      case e: Exception =>
        logError(s"MinIO test failed: ${e.getMessage}")
        logError(s"Stack trace: ${e.getStackTrace.mkString("\n")}")
        false
    }
  }

  def testHiveMetastore(spark: SparkSession): Boolean = {
    logInfo("\n=== Testing Hive Metastore ===")
    try {
      val df = createTestDataFrame(spark)
      val tableName = "test_hive_table"
      
      logInfo(s"Creating Hive table: $tableName")
      df.write
        .mode("overwrite")
        .saveAsTable(tableName)
      
      val readDf = spark.table(tableName)
      logInfo("Read DataFrame from Hive:")
      readDf.show()
      true
    } catch {
      case e: Exception =>
        logError(s"Hive test failed: ${e.getMessage}")
        logError(s"Stack trace: ${e.getStackTrace.mkString("\n")}")
        false
    }
  }

  def testDeltaLake(spark: SparkSession): Boolean = {
    logInfo("\n=== Testing Delta Lake ===")
    try {
      val df = createTestDataFrame(spark)
      val deltaPath = s"s3a://${sys.env("MINIO_BUCKET")}/delta/test_delta"
      
      logInfo(s"Creating Delta table at $deltaPath")
      df.write
        .format("delta")
        .mode("overwrite")
        .save(deltaPath)
      
      val readDf = spark.read
        .format("delta")
        .load(deltaPath)
      
      logInfo("Initial Delta table:")
      readDf.show()
      
      val updatesDf = createTestDataFrame(spark)
      logInfo("Updating Delta table...")
      
      val deltaTable = DeltaTable.forPath(spark, deltaPath)
      deltaTable.alias("target")
        .merge(
          updatesDf.alias("source"),
          "target.id = source.id"
        )
        .whenMatched()
        .updateAll()
        .whenNotMatched()
        .insertAll()
        .execute()
      
      val updatedDf = spark.read
        .format("delta")
        .load(deltaPath)
      
      logInfo("Updated Delta table:")
      updatedDf.show()
      true
    } catch {
      case e: Exception =>
        logError(s"Delta Lake test failed: ${e.getMessage}")
        logError(s"Stack trace: ${e.getStackTrace.mkString("\n")}")
        false
    }
  }

  private def createTestDataFrame(spark: SparkSession): DataFrame = {
    val schema = StructType(Seq(
      StructField("id", IntegerType, nullable = false),
      StructField("name", StringType, nullable = false)
    ))
    
    val data = Seq(
      (1, "test1"),
      (2, "test2"),
      (3, "test3")
    )
    
    spark.createDataFrame(data).toDF("id", "name")
  }

  private def createMinioBucket(): Unit = {
    logInfo("\n=== Creating MinIO Bucket if it doesn't exist ===")
    val bucket = sys.env("MINIO_BUCKET")
    
    try {
      // Create a temporary script to create the bucket
      val script = s"""
        |#!/bin/bash
        |mc alias set myminio ${sys.env("AWS_ENDPOINT_URL")} ${sys.env("AWS_ACCESS_KEY_ID")} ${sys.env("AWS_SECRET_ACCESS_KEY")}
        |mc mb myminio/$bucket --ignore-existing
        |mc ls myminio/$bucket
        |""".stripMargin
      
      val scriptFile = new java.io.File("/tmp/create_bucket.sh")
      val writer = new java.io.FileWriter(scriptFile)
      writer.write(script)
      writer.close()
      scriptFile.setExecutable(true)
      
      // Execute the script
      val process = Runtime.getRuntime.exec(Array("/bin/bash", scriptFile.getAbsolutePath))
      val exitCode = process.waitFor()
      
      if (exitCode == 0) {
        logInfo(s"Successfully created/verified bucket: $bucket")
      } else {
        val error = scala.io.Source.fromInputStream(process.getErrorStream).mkString
        logError(s"Failed to create bucket: $error")
        throw new RuntimeException(s"Failed to create MinIO bucket: $bucket")
      }
    } catch {
      case e: Exception =>
        logError(s"Error creating MinIO bucket: ${e.getMessage}")
        throw e
    }
  }
} 

File: tests/test_spark_cluster.py
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, IntegerType
import os
import time

def create_spark_session():
    """Create a Spark session with all necessary configurations."""
    return (SparkSession.builder
            .appName("SparkClusterTest")
            .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
            .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
            .getOrCreate())

def test_basic_spark(spark):
    """Test basic Spark functionality."""
    print("\n=== Testing Basic Spark Functionality ===")
    
    # Create a simple DataFrame
    data = [("Alice", 1), ("Bob", 2), ("Charlie", 3)]
    schema = StructType([
        StructField("name", StringType(), True),
        StructField("age", IntegerType(), True)
    ])
    df = spark.createDataFrame(data, schema)
    
    # Test DataFrame operations
    print("Created DataFrame:")
    df.show()
    
    # Test SQL
    df.createOrReplaceTempView("people")
    result = spark.sql("SELECT * FROM people WHERE age > 1")
    print("\nSQL Query Result:")
    result.show()
    
    return True

def test_minio_connectivity(spark):
    """Test MinIO connectivity and basic S3 operations."""
    print("\n=== Testing MinIO Connectivity ===")
    
    # Create a test DataFrame
    data = [("test1", 1), ("test2", 2)]
    df = spark.createDataFrame(data, ["key", "value"])
    
    # Get bucket name from environment
    bucket = os.getenv("MINIO_BUCKET", "spark-data")
    test_path = f"s3a://{bucket}/test/minio_test"
    
    # Write to MinIO
    print(f"Writing test data to {test_path}")
    df.write.mode("overwrite").parquet(test_path)
    
    # Read from MinIO
    print("Reading test data from MinIO")
    read_df = spark.read.parquet(test_path)
    print("Data read from MinIO:")
    read_df.show()
    
    return True

def test_hive_metastore(spark):
    """Test Hive metastore connectivity and operations."""
    print("\n=== Testing Hive Metastore ===")
    
    # Create a test table
    table_name = "test_hive_table"
    data = [("hive1", 1), ("hive2", 2)]
    df = spark.createDataFrame(data, ["name", "value"])
    
    # Write to Hive
    print(f"Creating Hive table: {table_name}")
    df.write.mode("overwrite").saveAsTable(table_name)
    
    # Read from Hive
    print("Reading from Hive table")
    result = spark.sql(f"SELECT * FROM {table_name}")
    print("Data from Hive table:")
    result.show()
    
    return True

def test_delta_lake(spark):
    """Test Delta Lake functionality."""
    print("\n=== Testing Delta Lake ===")
    
    # Get bucket name from environment
    bucket = os.getenv("MINIO_BUCKET", "spark-data")
    delta_path = f"s3a://{bucket}/delta/test_delta"
    
    # Create initial data
    data = [("delta1", 1), ("delta2", 2)]
    df = spark.createDataFrame(data, ["name", "value"])
    
    # Write as Delta table
    print(f"Creating Delta table at {delta_path}")
    df.write.format("delta").mode("overwrite").save(delta_path)
    
    # Read Delta table
    print("Reading Delta table")
    delta_df = spark.read.format("delta").load(delta_path)
    print("Initial Delta table data:")
    delta_df.show()
    
    # Perform an update
    print("Performing Delta table update")
    new_data = [("delta3", 3), ("delta4", 4)]
    update_df = spark.createDataFrame(new_data, ["name", "value"])
    update_df.write.format("delta").mode("append").save(delta_path)
    
    # Read updated data
    print("Reading updated Delta table")
    updated_df = spark.read.format("delta").load(delta_path)
    print("Updated Delta table data:")
    updated_df.show()
    
    return True

def main():
    """Run all tests."""
    print("Starting Spark Cluster Tests...")
    
    # Create Spark session
    spark = create_spark_session()
    
    try:
        # Run all tests
        tests = [
            ("Basic Spark", test_basic_spark),
            ("MinIO", test_minio_connectivity),
            ("Hive", test_hive_metastore),
            ("Delta Lake", test_delta_lake)
        ]
        
        results = []
        for test_name, test_func in tests:
            print(f"\nRunning {test_name} test...")
            try:
                result = test_func(spark)
                results.append((test_name, result, None))
                print(f"✓ {test_name} test passed")
            except Exception as e:
                results.append((test_name, False, str(e)))
                print(f"✗ {test_name} test failed: {str(e)}")
        
        # Print summary
        print("\n=== Test Summary ===")
        all_passed = True
        for test_name, passed, error in results:
            status = "PASSED" if passed else f"FAILED: {error}"
            print(f"{test_name}: {status}")
            if not passed:
                all_passed = False
        
        return 0 if all_passed else 1
        
    finally:
        spark.stop()

if __name__ == "__main__":
    exit(main()) 